oak_functions/loader/src/server.rs

```diff
@@ -149,25 +148,54 @@ impl FixedSizeBodyPadder for Response {
     }
 }
 
+/// Trait for implementing extensions, to implement new native functionality.
+pub trait OakApiNativeExtension {
+    /// Similar to `invoke_index` in [`wasmi::Externals`], but may return a result to be
+    /// written into the memory of the `WasmState`.
+    fn invoke_index(
+        &self,
+        wasm_state: &WasmState,
+        args: wasmi::RuntimeArgs,
+    ) -> Result<Result<Option<ExtensionResult>, OakStatus>, wasmi::Trap>;
+
+    /// Similar to `resolve_func` in [`wasmi::ModuleImportResolver`], but returns a tuple with a
+    /// function index and a function signature to be converted into [`wasmi::FuncRef`].
+    fn resolve_func(&self) -> anyhow::Result<(usize, wasmi::Signature)>;
+
+    /// Registration info, including a function index and a function name, for registering this
+    /// extension in the [`WasmState`].
+    fn registration_info(&self) -> (usize, String);
+}
+
+pub type BoxedExtension = Box<dyn OakApiNativeExtension + Send + Sync>;
+
+// Result of the invocation of an extension. The [`WasmState`] will write the `bytes` into the
+// memory location specified by `buf_ptr_ptr` and `buf_len_ptr`.
+pub struct ExtensionResult {
+    pub bytes: Vec<u8>,
+    pub buf_ptr_ptr: AbiPointer,
+    pub buf_len_ptr: AbiPointer,
+}
+
 /// `WasmState` holds runtime values for a particular execution instance of Wasm, handling a
 /// single user request. The methods here correspond to the ABI host functions that allow the Wasm
 /// module to exchange the request and the response with the Oak functions server. These functions
 /// translate values between Wasm linear memory and Rust types.
-struct WasmState {
+pub struct WasmState {
     request_bytes: Vec<u8>,
     response_bytes: Vec<u8>,
     lookup_data: Arc<LookupData>,
-    #[allow(dead_code)]
-    tf_model: Arc<Option<TensorFlowModel>>,
     instance: Option<wasmi::ModuleRef>,
     memory: Option<wasmi::MemoryRef>,
     logger: Logger,
     metrics_proxy: Option<PrivateMetricsProxy>,
+    extensions_indexes: Arc<RwLock<HashMap<usize, Arc<BoxedExtension>>>>,
+    extensions_names: Arc<RwLock<HashMap<String, Arc<BoxedExtension>>>>,
```

Comment:
Intuitively I would not expect all this synchronization around a hashmap that essentially never changes once created; why is it necessary?

If anything, the synchronization should be around each Extension, if a single instance is assumed to not be invoked concurrently.